### ISSUE TITLE
IO::Poll - undef fd not checked in mask()

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7525]
Migrated from [rt.perl.org#31864](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=31864) (status was 'resolved')

Searchable as RT31864$

**Comment by p5pRT** [comment]
### From Peter.Dintelmann@dresdner-bank.com

This is a bug report for perl from Peter\.Dintelmann@&#8203;dresdner\-bank\.com\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.5\.

When IO&#8203;::Poll cannot determine the file descriptor of the first
argument passed to the mask\(\) method\, STDIN is later polled instead\.
This seems to be due to the fact that the undef value returned by 
fileno\(\) evaluates to 0 in numeric context\.

Consider the following example where accidentally "F" is passed
to mask\(\) instead of "\\\*F"\.

&nbsp;    $ truss \-tpoll \-vpoll perl \-MIO&#8203;::Poll=POLLIN \-e '$p=IO&#8203;::Poll\->new\(\);
&nbsp;    open F\, "/dev/null" or die $\!; $p\->mask\(F=>POLLIN\); $p\->poll\(\)'
&nbsp;    \</dev/null
&nbsp;    poll\(0x00035210\, 1\, \-1\)                         = 1
&nbsp;            fd=0  ev=POLLIN rev=POLLIN
&nbsp;    $

The patch below should fix this\. Can you please let me know if it is
ok since this is my first patch ever?

TIA\,

&nbsp;    Peter

\-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm    Thu Sep  9
14&#8203;:45
:54 2004
\+\+\+ /var/tmp/Poll\.pm    Wed Oct  6 10&#8203;:30&#8203;:50 2004
@&#8203;@&#8203; \-8\,6 \+8\,7 @&#8203;@&#8203;
 package IO&#8203;::Poll;

 use strict;
\+use Carp;
 use IO&#8203;::Handle;
 use Exporter \(\);
 our\(@&#8203;ISA\, @&#8203;EXPORT\_OK\, @&#8203;EXPORT\, $VERSION\);
@&#8203;@&#8203; \-46\,6 \+47\,10 @&#8203;@&#8203;
&nbsp;     my $self = shift;
&nbsp;     my $io = shift;
&nbsp;     my $fd = fileno\($io\);
\+    unless \(defined $fd\) \{
\+        carp "Cannot determine the file descriptor of $io";
\+        return;
\+    \}
&nbsp;     if \(@&#8203;\_\) \{
&nbsp;        my $mask = shift;
&nbsp;        if\($mask\) \{



---
Flags&#8203;:
&nbsp;    category=library
&nbsp;    severity=low


---
Site configuration information for perl v5\.8\.5&#8203;:

Configured by p013176 at Thu Sep  9 14&#8203;:17&#8203;:03 CEST 2004\.

Summary of my perl5 \(revision 5 version 8 subversion 5\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.9\, archname=sun4\-solaris\-thread\-multi
&nbsp;    uname='sunos oikbm03 5\.9 generic\_112233\-01 sun4u sparc sunw\,ultra\-5\_10 '
&nbsp;    config\_args='\-Dprefix=/opt/perl \-Uinstallusrbinperl \-Dcc=gcc
\-Doptimize=\-O2
\-Duseshrplib \-Duseperlio \-Dusethreads \-des'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define
usemultiplicity=de
fine
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe
\-I/usr/local/inc
lude \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='3\.3\.2'\, gccosandvers='solaris2\.9'
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize
=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/local/lib '
&nbsp;    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-lgdbm \-ldl \-lm \-lpthread \-lc
&nbsp;    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lpthread \-lc
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='  \-R
/opt/perl/lib
/5\.8\.5/sun4\-solaris\-thread\-multi/CORE'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib'

Locally applied patches&#8203;:




---
@&#8203;INC for perl v5\.8\.5&#8203;:
&nbsp;    /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi
&nbsp;    /opt/perl/lib/5\.8\.5
&nbsp;    /opt/perl/lib/site\_perl/5\.8\.5/sun4\-solaris\-thread\-multi
&nbsp;    /opt/perl/lib/site\_perl/5\.8\.5
&nbsp;    /opt/perl/lib/site\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.5&#8203;:
&nbsp;    HOME=/export/home/p013176
&nbsp;    LANG \(unset\)
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/sbin/&#8203;:/opt/perl/bin/&#8203;:/usr/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/sh

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Peter \. Dintelmann @&#8203; dresdner\-bank \. com \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
>\# New Ticket Created by  Peter\.Dintelmann@&#8203;dresdner\-bank\.com 
>\# Please include the string&#8203;:  \[perl \#31864\]
>\# in the subject line of all future correspondence about this issue\. 
>\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=31864 >
>
>
>This is a bug report for perl from Peter\.Dintelmann@&#8203;dresdner\-bank\.com\,
>generated with the help of perlbug 1\.35 running under perl v5\.8\.5\.
>
>When IO&#8203;::Poll cannot determine the file descriptor of the first
>argument passed to the mask\(\) method\, STDIN is later polled instead\.
>This seems to be due to the fact that the undef value returned by 
>fileno\(\) evaluates to 0 in numeric context\.

I am not sure I like the carp\(\)\.
This should either be fatal or silent and return false\.


>
>Consider the following example where accidentally "F" is passed
>to mask\(\) instead of "\\\*F"\.
>
>    $ truss \-tpoll \-vpoll perl \-MIO&#8203;::Poll=POLLIN \-e '$p=IO&#8203;::Poll\->new\(\);
>    open F\, "/dev/null" or die $\!; $p\->mask\(F=>POLLIN\); $p\->poll\(\)'
>    \</dev/null
>    poll\(0x00035210\, 1\, \-1\)                         = 1
>            fd=0  ev=POLLIN rev=POLLIN
>    $
>
>The patch below should fix this\. Can you please let me know if it is
>ok since this is my first patch ever?
>
>TIA\,
>
>    Peter
>
>\-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm    Thu Sep  9
>14&#8203;:45
>&#8203;:54 2004
>\+\+\+ /var/tmp/Poll\.pm    Wed Oct  6 10&#8203;:30&#8203;:50 2004
>@&#8203;@&#8203; \-8\,6 \+8\,7 @&#8203;@&#8203;
> package IO&#8203;::Poll;
>
> use strict;
>\+use Carp;
> use IO&#8203;::Handle;
> use Exporter \(\);
> our\(@&#8203;ISA\, @&#8203;EXPORT\_OK\, @&#8203;EXPORT\, $VERSION\);
>@&#8203;@&#8203; \-46\,6 \+47\,10 @&#8203;@&#8203;
>     my $self = shift;
>     my $io = shift;
>     my $fd = fileno\($io\);
>\+    unless \(defined $fd\) \{
>\+        carp "Cannot determine the file descriptor of $io";
>\+        return;
>\+    \}
>     if \(@&#8203;\_\) \{
>        my $mask = shift;
>        if\($mask\) \{
>
>\-\-\-
>Flags&#8203;:
>    category=library
>    severity=low
>\-\-\-
>Site configuration information for perl v5\.8\.5&#8203;:
>
>Configured by p013176 at Thu Sep  9 14&#8203;:17&#8203;:03 CEST 2004\.
>
>Summary of my perl5 \(revision 5 version 8 subversion 5\) configuration&#8203;:
>  Platform&#8203;:
>    osname=solaris\, osvers=2\.9\, archname=sun4\-solaris\-thread\-multi
>    uname='sunos oikbm03 5\.9 generic\_112233\-01 sun4u sparc sunw\,ultra\-5\_10 '
>    config\_args='\-Dprefix=/opt/perl \-Uinstallusrbinperl \-Dcc=gcc
>\-Doptimize=\-O2
>\-Duseshrplib \-Duseperlio \-Dusethreads \-des'
>    hint=recommended\, useposix=true\, d\_sigaction=define
>    usethreads=define use5005threads=undef useithreads=define
>usemultiplicity=de
>fine
>    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
>    use64bitint=undef use64bitall=undef uselongdouble=undef
>    usemymalloc=n\, bincompat5005=undef
>  Compiler&#8203;:
>    cc='gcc'\, ccflags ='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe
>\-I/usr/local/inc
>lude \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
>    optimize='\-O2'\,
>    cppflags='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe \-I/usr/local/include'
>    ccversion=''\, gccversion='3\.3\.2'\, gccosandvers='solaris2\.9'
>    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
>    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
>    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
>lseeksize
>=8
>    alignbytes=8\, prototype=define
>  Linker and Libraries&#8203;:
>    ld='gcc'\, ldflags =' \-L/usr/local/lib '
>    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
>    libs=\-lsocket \-lnsl \-lgdbm \-ldl \-lm \-lpthread \-lc
>    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lpthread \-lc
>    libc=/lib/libc\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
>    gnulibc\_version=''
>  Dynamic Linking&#8203;:
>    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='  \-R
>/opt/perl/lib
>/5\.8\.5/sun4\-solaris\-thread\-multi/CORE'
>    cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib'
>
>Locally applied patches&#8203;:
>
>
>\-\-\-
>@&#8203;INC for perl v5\.8\.5&#8203;:
>    /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi
>    /opt/perl/lib/5\.8\.5
>    /opt/perl/lib/site\_perl/5\.8\.5/sun4\-solaris\-thread\-multi
>    /opt/perl/lib/site\_perl/5\.8\.5
>    /opt/perl/lib/site\_perl
>    \.
>
>\-\-\-
>Environment for perl v5\.8\.5&#8203;:
>    HOME=/export/home/p013176
>    LANG \(unset\)
>    LANGUAGE \(unset\)
>    LD\_LIBRARY\_PATH \(unset\)
>    LOGDIR \(unset\)
>    PATH=/usr/sbin/&#8203;:/opt/perl/bin/&#8203;:/usr/bin&#8203;:/usr/local/bin
>    PERL\_BADLANG \(unset\)
>    SHELL=/bin/sh

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From Peter.Dintelmann@dresdner-bank.com

&nbsp;    \[ Sorry for the late reply; I did not have access to
&nbsp;    my email for a few days \]

> \-\-\-\-\-Ursprüngliche Nachricht\-\-\-\-\-
> Von&#8203;: Nick Ing\-Simmons \[mailto&#8203;:nick@&#8203;ing\-simmons\.net\]
> Gesendet&#8203;: Freitag\, 8\. Oktober 2004 11&#8203;:39
> An&#8203;: perl5\-porters@&#8203;perl\.org
> Cc&#8203;: bugs\-bitbucket@&#8203;rt\.perl\.org
> Betreff&#8203;: Re&#8203;: \[perl \#31864\] IO&#8203;::Poll \- undef fd not checked in mask\(\)
> 
> 
> Peter \. Dintelmann @&#8203; dresdner\-bank \. com 
> \<perl5\-porters@&#8203;perl\.org> writes&#8203;:
> >\# New Ticket Created by  Peter\.Dintelmann@&#8203;dresdner\-bank\.com 
> >\# Please include the string&#8203;:  \[perl \#31864\]
> >\# in the subject line of all future correspondence about this issue\. 
> >\# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=31864 >
> >
> >
> >This is a bug report for perl from 
> Peter\.Dintelmann@&#8203;dresdner\-bank\.com\,
> >generated with the help of perlbug 1\.35 running under perl v5\.8\.5\.
> >
> >When IO&#8203;::Poll cannot determine the file descriptor of the first
> >argument passed to the mask\(\) method\, STDIN is later polled instead\.
> >This seems to be due to the fact that the undef value returned by 
> >fileno\(\) evaluates to 0 in numeric context\.
> 
> I am not sure I like the carp\(\)\.
> This should either be fatal or silent and return false\.


&nbsp;    Personally I prefer the silent mode\. 

\-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm    Thu Sep  9
14&#8203;:45&#8203;:54 2004
\+\+\+ /var/tmp/Poll\.pm    Mon Oct 11 09&#8203;:49&#8203;:48 2004
@&#8203;@&#8203; \-46\,6 \+46\,7 @&#8203;@&#8203;
&nbsp;     my $self = shift;
&nbsp;     my $io = shift;
&nbsp;     my $fd = fileno\($io\);
\+    return unless defined $fd;
&nbsp;     if \(@&#8203;\_\) \{
&nbsp;        my $mask = shift;
&nbsp;        if\($mask\) \{

&nbsp;    Is this ok?


> >Consider the following example where accidentally "F" is passed
> >to mask\(\) instead of "\\\*F"\.
> >
> >    $ truss \-tpoll \-vpoll perl \-MIO&#8203;::Poll=POLLIN \-e 
> '$p=IO&#8203;::Poll\->new\(\);
> >    open F\, "/dev/null" or die $\!; $p\->mask\(F=>POLLIN\); $p\->poll\(\)'
> >    \</dev/null
> >    poll\(0x00035210\, 1\, \-1\)                         = 1
> >            fd=0  ev=POLLIN rev=POLLIN
> >    $
> >
> >The patch below should fix this\. Can you please let me know if it is
> >ok since this is my first patch ever?
> >
> >TIA\,
> >
> >    Peter
> >
> >\-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm 
>    Thu Sep  9
> >14&#8203;:45
> >&#8203;:54 2004
> >\+\+\+ /var/tmp/Poll\.pm    Wed Oct  6 10&#8203;:30&#8203;:50 2004
> >@&#8203;@&#8203; \-8\,6 \+8\,7 @&#8203;@&#8203;
> > package IO&#8203;::Poll;
> >
> > use strict;
> >\+use Carp;
> > use IO&#8203;::Handle;
> > use Exporter \(\);
> > our\(@&#8203;ISA\, @&#8203;EXPORT\_OK\, @&#8203;EXPORT\, $VERSION\);
> >@&#8203;@&#8203; \-46\,6 \+47\,10 @&#8203;@&#8203;
> >     my $self = shift;
> >     my $io = shift;
> >     my $fd = fileno\($io\);
> >\+    unless \(defined $fd\) \{
> >\+        carp "Cannot determine the file descriptor of $io";
> >\+        return;
> >\+    \}
> >     if \(@&#8203;\_\) \{
> >        my $mask = shift;
> >        if\($mask\) \{
> >
> >\-\-\-
> >Flags&#8203;:
> >    category=library
> >    severity=low
> >\-\-\-
> >Site configuration information for perl v5\.8\.5&#8203;:
> >
> >Configured by p013176 at Thu Sep  9 14&#8203;:17&#8203;:03 CEST 2004\.
> >
> >Summary of my perl5 \(revision 5 version 8 subversion 5\) 
> configuration&#8203;:
> >  Platform&#8203;:
> >    osname=solaris\, osvers=2\.9\, archname=sun4\-solaris\-thread\-multi
> >    uname='sunos oikbm03 5\.9 generic\_112233\-01 sun4u sparc 
> sunw\,ultra\-5\_10 '
> >    config\_args='\-Dprefix=/opt/perl \-Uinstallusrbinperl \-Dcc=gcc
> >\-Doptimize=\-O2
> >\-Duseshrplib \-Duseperlio \-Dusethreads \-des'
> >    hint=recommended\, useposix=true\, d\_sigaction=define
> >    usethreads=define use5005threads=undef useithreads=define
> >usemultiplicity=de
> >fine
> >    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
> >    use64bitint=undef use64bitall=undef uselongdouble=undef
> >    usemymalloc=n\, bincompat5005=undef
> >  Compiler&#8203;:
> >    cc='gcc'\, ccflags ='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe
> >\-I/usr/local/inc
> >lude \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
> >    optimize='\-O2'\,
> >    cppflags='\-D\_REENTRANT \-fno\-strict\-aliasing \-pipe 
> \-I/usr/local/include'
> >    ccversion=''\, gccversion='3\.3\.2'\, gccosandvers='solaris2\.9'
> >    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
> >    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, 
> longdblsize=16
> >    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, 
> Off\_t='off\_t'\,
> >lseeksize
> >=8
> >    alignbytes=8\, prototype=define
> >  Linker and Libraries&#8203;:
> >    ld='gcc'\, ldflags =' \-L/usr/local/lib '
> >    libpth=/usr/local/lib /usr/lib /usr/ccs/lib
> >    libs=\-lsocket \-lnsl \-lgdbm \-ldl \-lm \-lpthread \-lc
> >    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lpthread \-lc
> >    libc=/lib/libc\.so\, so=so\, useshrplib=true\, libperl=libperl\.so
> >    gnulibc\_version=''
> >  Dynamic Linking&#8203;:
> >    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='  \-R
> >/opt/perl/lib
> >/5\.8\.5/sun4\-solaris\-thread\-multi/CORE'
> >    cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib'
> >
> >Locally applied patches&#8203;:
> >
> >
> >\-\-\-
> >@&#8203;INC for perl v5\.8\.5&#8203;:
> >    /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi
> >    /opt/perl/lib/5\.8\.5
> >    /opt/perl/lib/site\_perl/5\.8\.5/sun4\-solaris\-thread\-multi
> >    /opt/perl/lib/site\_perl/5\.8\.5
> >    /opt/perl/lib/site\_perl
> >    \.
> >
> >\-\-\-
> >Environment for perl v5\.8\.5&#8203;:
> >    HOME=/export/home/p013176
> >    LANG \(unset\)
> >    LANGUAGE \(unset\)
> >    LD\_LIBRARY\_PATH \(unset\)
> >    LOGDIR \(unset\)
> >    PATH=/usr/sbin/&#8203;:/opt/perl/bin/&#8203;:/usr/bin&#8203;:/usr/local/bin
> >    PERL\_BADLANG \(unset\)
> >    SHELL=/bin/sh
>

**Comment by p5pRT** [comment]
### From @rgs

Dintelmann\, Peter wrote&#8203;:
> > I am not sure I like the carp\(\)\.
> > This should either be fatal or silent and return false\.
> 
> 
>     Personally I prefer the silent mode\. 
> 
> \-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm    Thu Sep  9
> 14&#8203;:45&#8203;:54 2004
> \+\+\+ /var/tmp/Poll\.pm    Mon Oct 11 09&#8203;:49&#8203;:48 2004

Thanks\, applied as change \#23366 to bleadperl\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From perl5-porters@ton.iguana.be

In article \<20041014181157\.2bb866a8@&#8203;valis\.local>\,
&nbsp;    Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;mandrakesoft\.com> writes&#8203;:
> Dintelmann\, Peter wrote&#8203;:
>> > I am not sure I like the carp\(\)\.
>> > This should either be fatal or silent and return false\.
>> 
>> 
>>     Personally I prefer the silent mode\. 
>> 
>> \-\-\- /opt/perl/lib/5\.8\.5/sun4\-solaris\-thread\-multi/IO/Poll\.pm    Thu Sep  9
>> 14&#8203;:45&#8203;:54 2004
>> \+\+\+ /var/tmp/Poll\.pm    Mon Oct 11 09&#8203;:49&#8203;:48 2004
> 
> Thanks\, applied as change \#23366 to bleadperl\.
May I pipe up that I dislike silent mode very much\. This is almost certainly
a user error perl is detecting here and then ignoring instead of helping\.
I think a warn is actually apporpiate since it's what the perl builtins do&#8203;:

perl \-wle 'sysread\(undef\, $a\, 4096\); print "foo"'
Name "main&#8203;::a" used only once&#8203;: possible typo at \-e line 1\.
Use of uninitialized value in ref\-to\-glob cast at \-e line 1\.
sysread\(\) on unopened filehandle at \-e line 1\.
foo
